### PR TITLE
Store: Fix panic too smaller buffer

### PR DESCRIPTION
**Pull request by dominicqi**
* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

Related issue  https://github.com/thanos-io/thanos/issues/6934
I think the problem could be: a buffer that was not obtained from the buffer pool should not be put back into it, as it will lead to buffers with incorrect sizes.

### CODE REVIEW

**Comment by saswatamcode** [comment]
Could you sign this commit?

**Comment by dominicqi** [comment]
> Could you sign this commit?

@saswatamcode signed, but the e2e test failed , i can't see  error information

**Review by GiedriusS — COMMENTED** [review]
Could you add a unit test for this?

**Comment by dominicqi** [comment]
> Could you add a unit test for this?

Hi, @GiedriusS  i haven't thought of what unit tests should be added yet. What do you think should be added? just moved the defer statement to the appropriate place.

**Comment by GiedriusS** [comment]
> > Could you add a unit test for this?
> 
> Hi, @GiedriusS i haven't thought of what unit tests should be added yet. What do you think should be added? just moved the defer statement to the appropriate place.

I think a test should call that function twice and in the second call it should get a wrongly sized buffer with this bug, right?

**Comment by dominicqi** [comment]
> > > Could you add a unit test for this?
> > 
> > 
> > Hi, @GiedriusS i haven't thought of what unit tests should be added yet. What do you think should be added? just moved the defer statement to the appropriate place.
> 
> I think a test should call that function twice and in the second call it should get a wrongly sized buffer with this bug, right?

I don't think so; the buffer pool returns a wrongly sized buffer because it has a smaller buffer. Just make sure only the obtained buffer can return to the pool.

**Comment by yeya24** [comment]
I think a unit test is still good to have to catch such regression next time.

File: pkg/store/bucket.go [hunk 1]
```diff
@@ -3577,10 +3577,10 @@ func (r *bucketChunkReader) loadChunks(ctx context.Context, res []seriesEntry, a
 	bufPooled, err := r.block.chunkPool.Get(r.block.estimatedMaxChunkSize)
 	if err == nil {
 		buf = *bufPooled
+		defer r.block.chunkPool.Put(&buf)
 	} else {
 		buf = make([]byte, r.block.estimatedMaxChunkSize)
```
Comment (yeya24) [inline pkg/store/bucket.go:3582]: Should we still put but back to the pool but extend buffer size to the bucket size?

Comment (dominicqi) [reply pkg/store/bucket.go:3582]: The buffer is only created when `ErrPoolExhausted` occurs, it should not be put back.

**Comment by dominicqi** [comment]
> I think a unit test is still good to have to catch such regression next time.

**Comment by yeya24** [comment]
I will get this merged to fix the bug. Thanks @dominicqi.

**Review by yeya24 — APPROVED** [review]

**Pull request merged**